iothub/device/src/Transport/Amqp/AmqpUnit.cs

```diff
@@ -293,95 +328,109 @@ public void OnEventsReceived(Message message)
-                Task<AmqpIoTSendingLink> sendingLinkCreator = _amqpIoTSession.OpenMethodsSenderLinkAsync(_deviceIdentity, correlationIdSuffix, timeout);
-                await Task.WhenAll(receiveLinkCreator, sendingLinkCreator).ConfigureAwait(false);
-
-                _methodReceivingLink = receiveLinkCreator.Result;
-                _methodSendingLink = sendingLinkCreator.Result;
-
-                _methodReceivingLink.RegisterMethodListener(OnMethodReceived);
-                _methodSendingLink.Closed += OnLinkDisconnected;
-                _methodReceivingLink.Closed += OnLinkDisconnected;
+                throw DEVICE_DISCONNECTED_EXCEPTION;
+            }
 
-                if (Logging.IsEnabled) Logging.Associate(this, _methodReceivingLink, $"{nameof(_methodReceivingLink)}");
-                if (Logging.IsEnabled) Logging.Associate(this, _methodSendingLink, $"{nameof(_methodSendingLink)}");
+            if (Logging.IsEnabled) Logging.Enter(this, timeout, $"{nameof(EnableMethodsAsync)}");
+            AmqpIoTSession amqpIoTSession = await EnsureSessionAsync(timeout).ConfigureAwait(false);
+            bool gain = await _methodLinkLock.WaitAsync(timeout).ConfigureAwait(false);
+            if (!gain)
+            {
+                throw new TimeoutException();
             }
-            catch (Exception)
+            
+            string correlationIdSuffix = Guid.NewGuid().ToString();
+            try
             {
-                _methodReceivingLink?.Abort();
-                _methodReceivingLink = null;
-
-                _methodSendingLink?.Abort();
-                _methodSendingLink = null;
-
-                throw;
+                await Task.WhenAll(
+                    OpenMethodsReceiverLinkAsync(amqpIoTSession, correlationIdSuffix, timeout),
+                    OpenMethodsSenderLinkAsync(amqpIoTSession, correlationIdSuffix, timeout)
+                ).ConfigureAwait(false);
             }
             finally
             {
+                _methodLinkLock.Release();
                 if (Logging.IsEnabled) Logging.Exit(this, timeout, $"{nameof(EnableMethodsAsync)}");
             }
         }
-
-        private void OnMethodReceived(MethodRequestInternal methodRequestInternal)
+        
+        private async Task OpenMethodsReceiverLinkAsync(AmqpIoTSession amqpIoTSession, string correlationIdSuffix, TimeSpan timeout)
         {
-            if (Logging.IsEnabled) Logging.Enter(this, methodRequestInternal, $"{nameof(OnMethodReceived)}");
-            try
+            if (_methodReceivingLink == null || _methodReceivingLink.IsClosing())
             {
-                _methodHandler?.Invoke(methodRequestInternal);
+                _methodReceivingLink = await amqpIoTSession.OpenMethodsReceiverLinkAsync(_deviceIdentity, correlationIdSuffix, timeout).ConfigureAwait(false);
+                _methodReceivingLink.Closed += (obj, arg) => {
```

Comment:
The callbacks need to be configured before attaching the link to ensure that it's going to be called.
This code has a race condition as between line 363 and 364 the session can get closed and this function will use a closed link.

Reply:
Using a closed link is by design because when a closed link is used, it will throw an exception. In the next execution, it will try to ensure that it is opened again.